packages/appinio_swiper/lib/appinio_swiper.dart

```diff
@@ -1,679 +1,743 @@
-              ? AppinioSwiperDirection.left
-              : AppinioSwiperDirection.bottom);
-    } else {
-      direction = top < 0
-          ? (left < top.abs()
-              ? AppinioSwiperDirection.top
-              : AppinioSwiperDirection.right)
-          : (left > top
-              ? AppinioSwiperDirection.right
-              : AppinioSwiperDirection.bottom);
+  int _effectiveBackgroundCardCount(int index) {
+    if (widget.cardCount == 0) {
+      return 0;
+    }
+    // Use one extra card so cards entering the stack can fade in smoothly.
+    final int effectiveCardCount = widget.backgroundCardCount + 1;
+    if (widget.loop) {
+      return effectiveCardCount;
     }
-    return direction;
+    final int remaining = widget.cardCount - position.index - 1;
+    return (_swipeActivity is Unswipe ? remaining + 1 : remaining)
+        .clamp(0, effectiveCardCount);
   }
 
   Future<void> _onSwiping() async {
-    widget.onSwiping?.call(_calculateDirection(top: _top, left: _left));
+    widget.onCardPositionChanged?.call(
+      position.offset.toAxisDirection(),
+      position.offset,
+    );
   }
 
-  void _calculateAngle() {
-    if (_angle <= _maxAngle && _angle >= -_maxAngle) {
-      (_tapOnTop == true)
-          ? _angle = (_maxAngle / 100) * (_left / 10)
-          : _angle = (_maxAngle / 100) * (_left / 10) * -1;
-    }
+  Offset _directionToTarget(AxisDirection direction) {
+    final Size size = MediaQuery.sizeOf(context);
+    return switch (direction) {
+      AxisDirection.up => Offset(0, -size.height),
+      AxisDirection.down => Offset(0, size.height),
+      AxisDirection.left => Offset(-size.width, 0),
+      AxisDirection.right => Offset(size.width, 0),
+    };
   }
 
-  void _calculateScale() {
-    if (_scale <= 1.0 && _scale >= 0.9) {
-      _scale =
-          (_total > 0) ? 0.9 + (_total / 5000) : 0.9 + -1 * (_total / 5000);
+  Future<void> _onPanEnd() async {
+    // TODO: Use a ballistic simulation to determine if the swipe should be
+    // triggered or not.
+    // See the snapping behavior from `DraggableScrollableSheet`.
+    if (position.offsetRelativeToThreshold.dx.abs() < 1 &&
+        position.offsetRelativeToThreshold.dy.abs() < 1) {
+      return _onSwipeCancelled(context);
     }
+    await _onSwipe(position.offset.toAxisDirection());
   }
 
-  void _calculateDifference() {
-    if (_difference >= 0 && _difference <= _difference) {
-      _difference = (_total > 0)
-          ? widget.cardsSpacing - (_total / 10)
-          : widget.cardsSpacing + (_total / 10);
-    }
+  Future<void> _onSwipeDefault() async {
+    return _onSwipe(widget.defaultDirection);
   }
+}
 
-  void _onEndAnimation() {
-    if (_left < -widget.threshold || _left > widget.threshold) {
-      _swipeHorizontal(context);
-    } else if (_top < -widget.threshold || _top > widget.threshold) {
-      _swipeVertical(context);
-    } else {
-      _goBack(context);
-    }
+class _BackgroundCards extends StatelessWidget {
```

Comment:
While helper methods that return widgets are convenient, they can have some performance foot guns.
https://www.youtube.com/watch?v=IOyq-eTRhvo

It's best to just always use a sub-widget rather than to navigate the foot guns.